packages/query-core/src/types.ts

```diff
@@ -200,9 +200,13 @@ export interface QueryObserverOptions<
    * If set, the component will only re-render if any of the listed properties change.
    * When set to `['data', 'error']`, the component will only re-render when the `data` or `error` properties change.
    * When set to `'all'`, the component will re-render whenever a query is updated.
+   * When set to a function, the function will be executed to compute the list of properties.
    * By default, access to properties will be tracked, and the component will only re-render when one of the tracked properties change.
    */
-  notifyOnChangeProps?: Array<keyof InfiniteQueryObserverResult> | 'all'
+  notifyOnChangeProps?:
+    | Array<keyof InfiniteQueryObserverResult>
+    | 'all'
+    | (() => Array<keyof InfiniteQueryObserverResult> | 'all')
```

Comment:
@TkDodo quick question - it would be totally fine to pass in a different function reference on every re-render, right?

I believe it would be same thing as passing in a new reference of `['data', 'error']`. 

```tsx
const { dataUpdatedAt } = useQuery({
    queryKey: ['myExampleKey'],
    queryFn: async () => {
      const response = await fetch('https://api.github.com/repos/tannerlinsley/react-query');
      return response.json();
    },
    notifyOnChangeProps: ['data', 'error']
  });
```

Reply:
yes, referential stability doesn't matter here.